### PR TITLE
Fixing possible memory leak when break is used with for await of

### LINKED ISSUE #13: Enhamcement/Idea: cancelable/abortable iterators
I was thinking about a problem we can have with this library due to the iterator design:
imagine a code like this:
```ts
async function* firstIteration(): AsyncIterable<number> {
  const iterable = forEmitOf(someHeavyEmitter);
  for await (const item of iterable) {
    yield item * 2;
  }
}

async function* secondIteration(previousIterable: : AsyncIterable<number>) {
  for await (const item of previousIterable) {
      if (item > 20) {
          break;
      }
      yield item;
  }
}

async function run() {
   for await (const item of secondIteration(firstIteration())) {
    console.log(item);
   }
}

run();
```

The **break** on the second iterable, will make the code behave apparently as expected, but, behind the curtains, the events array created by the **forEmitOf** may keep getting new items, which can lead, ultimately, for serious memory leaks.

### Idea

As a matter of fact, the user of this library must be aware of this situation to not **break** the for await in cases like that, but it would be nice to have some aborting mechanism to make sure no hidden process will keep filling the memory. But such a mechanism must be propagable, as I can link an indefinite number of iterables. Also, I think it should be separated from this library, as it is not necessarily linked to iterables generated from EventEmitters.

The first idea was something used like that:
```ts
function firstIteration(): AsyncIterable<number> {
  return cancellable(async function* () {
    // internally, forEmitOf would return a cancellable iterable with a custom operation
    const iterable = forEmitOf(someHeavyEmitter); 
    for await (const item of iterable) {
      yield item * 2;
    }
  });
}

function secondIteration(previousIterable: : AsyncIterable<number>) {
  return cancellable(async function* internal() {
    for await (const item of previousIterable) {
        if (item > 20) {
            cancel(iterable);
            break;
        }
        yield item;
    }
  });
}

async function run() {
   for await (const item of secondIteration(firstIteration())) {
    console.log(item);
   }
}

run();
```

While the iteration works from inside out (starting at the first iterable in the chain) the **cancel** call would work from out to inside, propagating the cancellation while the referenced iterable is cancellable.

I know it's a bit of a pain to use a feature designed like that, but it's the best idea I could figure out yet. In the library I'm creating I would make all this automatically, so, the usability would be not bad at all, but still, if the user uses a **for await of** with a **break** without treating the cancelation, he can fall in that issue.

What do you think? I really want some opinions about this approach because I'm not quite sure, and I feel that, if I want to publish [this freature](https://github.com/Codibre/fluent-iterable/pull/4) I need to deal with this situation, as it's quite easy for the user to lose itself in a chain of operations

### PR DESCRIPTION
**Pull request by Farenheith**
Closes #13 

After I opened the last issue I started to study the possibilities to solve this problem and I realized that the native iterable mechanic already has a way to treat that situation: the **return** method.

You can check documentation [here](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Generator/return)

The return method, when called, will complete the iterable, so, you can use it to run a custom code where you clean everything you need to clean. The **for await of** always calls the return method. The only problem is that, with the generator syntax, you can't (or at least I couldn't find a way to) implement it.

So, if you want to use it, unfortunately, you need to take the raw approach described [here](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Guide/Iteratores_e_geradores)

To make sure the code is working, I installed **sinon** and **chai** libraries (as dev dependencies) and I checked if some debugging methods are called as expected. This is a good way, I think, to check if the iteration runs as we expect behind the curtains :)
The code that checks it is this one, in the test case:
```ts
expect(debugging.debugYielding).to.have.been.calledTwice;
expect(debugging.debugIteratorReturn).to.have.been.called;
```
 Look that, if the function **debugYielding** is called just two times, that means the yielding happened just two times as well.
If the **debugIteratorReturn** is called, then it's guaranteed that the return method was called.

Another observation is that, with the new implementation, the generated iterable may support only one iteration, being needed to generate another one if you want to run two iterations. This is true, but that is the previous behavior as well, as you can see in this [playground](https://playcode.io/649082/).

I'll mark this PR as a draft because I'm not really sure if you want to let your code to follow this way. A raw iterable is harder to maintain, but maybe this approach will bring better benefits for your package overall.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -44,7 +46,12 @@
     "nodemon": "^2.0.4",
     "nyc": "^15.1.0",
     "prettier": "^2.0.5",
+    "sinon": "^9.0.2",
+    "sinon-chai": "^3.5.0",
     "ts-node": "^8.10.2",
     "typescript": "^3.9.7"
+  },
+  "dependencies": {
+    "abortable-iterator": "^3.0.0"
```
Comment (danstarns) [inline package.json]: Are we using `abortable-iterator`? I can't find it in the code, is it a peer of something?

**Comment by danstarns** [comment]
Firstly thank you for taking the time to look into this, I've been playing around & can't seem to find any breaking changes. I completely understand why you have chosen to not use `function *` syntax & if not using this will make the package better off I'm all for it! I see this is still in draft, I'm happy for it to go in as long as you are :)

File: package.json (hunk 1 above)
Comment (Farenheith) [reply package.json]: Thanks for notice it. I tried to use it first to solve the problem, until I realized I could solve it in a better way just taking advantage of the native iterator contract

**Comment by Farenheith** [comment]
@danstarns  Thank you for taking the time to analyze this PR! I removed the unused dependency and made it ready to review :)

**Pull request merged**